x-pack/plugins/lens/public/datatable_visualization/utils.ts

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { Datatable } from 'src/plugins/expressions/public';
+import { getOriginalId } from './transpose_helpers';
+
+function isValidNumber(value: unknown): boolean {
+  return typeof value === 'number' || value == null;
+}
+
+/**
+ * It checks the configuration and content of the current datatable
+ * It returns an object with the following content:
+ * * isNumeric: whether the field is numeric
+ * * hasNumberValues: whether the table rows contain all values of type number (empty values are ignored)
+ * * hasNumericValues: whether the table rows contain numeric compatible values (arrays with numbers are considered valid here, empty values are ignored)
+ * @param currentData
+ * @param accessor
+ * @returns An object containing stats metadata retrieved from both configuration and current data
+ */
+export function isNumericField(currentData: Datatable | undefined, accessor: string) {
+  const isNumeric =
+    currentData?.columns.find((col) => col.id === accessor || getOriginalId(col.id) === accessor)
+      ?.meta.type === 'number';
+
+  let hasFieldOnlyNumberValues = isNumeric;
+  let hasFieldNumericValues = isNumeric;
+  for (const row of currentData?.rows || []) {
+    const value = row[accessor];
+    hasFieldOnlyNumberValues = hasFieldOnlyNumberValues && isValidNumber(value);
+    hasFieldNumericValues =
+      (hasFieldNumericValues && hasFieldOnlyNumberValues) ||
```

Comment:
This logic is not correct - if there are arrays in the current column and the last row doesn't have a value, `hasFieldNumericValues` will become false: (first `FEMALE` column doesn't have a summary because the last row is undefined)
<img width="1593" alt="Screenshot 2021-06-07 at 14 18 38" src="https://user-images.githubusercontent.com/1508364/121015419-4536a700-c79b-11eb-89be-f839b7f81b32.png">

Also I found this logic extremely hard to follow and it seems like it's not used anywhere - `hasNumberValues` is never accessed and it seems like this function is only returning a complex object because of this and can be simplified to this for all consumers:
```ts
export function isNumericField(currentData: Datatable | undefined, accessor: string) {
  return currentData?.rows.every(row => {
    const val = row[accessor];
    return isValidNumber(val) || Array.isArray(val) && val.every(isValidNumber);
  });
}
```

Reply:
The specific distinction in this function was propaedeutical for more than this specific case, that's the reason it is in the root folder.
I've built on top of this PR a fix for dynamic coloring using this utility.

I've reworked it a bit to fix this specific issue and speed up the computation in case it's not valid. In case you think it's best to have something lighter and more feature specific I can rewrite it as well.

Reply:
If we don't know whether we are ever going to need it, please let's remove it for now. I only found this issue by chance because the data I was testing with had that specific case.